types/styled-system/index.d.ts

```diff
@@ -13,50 +13,63 @@
 //                 Chris LoPresto <https://github.com/chrislopresto>
 //                 Pedro Duarte <https://github.com/peduarte>
 //                 Dhalton Huber <https://github.com/Dhalton>
+//                 Elliot Bonneville <https://github.com/elliotbonneville>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.8
 
 import * as CSS from "csstype";
 
-export const defaultBreakpoints: string[];
-
-export function cloneFunc(fn: (...args: any[]) => any): (...args: any[]) => any;
 export function get(obj: any, ...paths: Array<string | number>): any;
-export function themeGet(keys: string, fallback?: string): any;
 
-export function is(n: any): boolean;
-export function isObject(n: any): boolean;
-export function num(n: any): boolean;
-export function px(n: any): string;
+export type Scale = object | Array<string | number>;
 
-export function createMediaQuery(n: number | string): string;
+// Preserved to support v4 shim:
+// https://github.com/styled-system/styled-system/blob/master/packages/styled-system/src/index.js#L108
+export interface LowLevelStyleFunctionArguments<N, S> {
+    prop: string;
+    cssProperty?: string;
+    alias?: string;
+    key?: string;
+    transformValue?: (n: N, scale?: S) => any;
+    scale?: S;
+    // new v5 api
+    properties?: string[];
+}
+
+export function style<N = string | number, S = Scale>(
+    // tslint:disable-next-line no-unnecessary-generics
+    args: LowLevelStyleFunctionArguments<N, S>
+): {
+    [cssProp: string]: string;
+};
 
 export interface styleFn {
     (...args: any[]): any;
-    propTypes?: string[];
 }
 
-export type Scale = object | Array<string | number>;
-
-export interface LowLevelStylefunctionArguments<N, S> {
-  prop: string;
-  cssProperty?: string;
-  alias?: string;
-  key?: string;
-  transformValue?: (n: N, scale?: S) => any;
-  scale?: S;
+export interface CustomStyleDefinition {
+    /** The CSS property to use in the returned style object (overridden by `properties` if present). */
+    property?: string;
+    /**
+     * An array of multiple properties (e.g. `['marginLeft', 'marginRight']`) to which this style's value will be
+     * assigned (overrides `property` when present).
+     */
+    properties?: string[];
+    /** A string referencing a key in the `theme` object. */
+    scale?: string;
+    /** A fallback scale object for when there isn't one defined in the `theme` object. */
+    defaultScale?: any[];
```

Comment:
Shouldn't this be a number's array?

Reply:
I think we should have a scale type

Reply:
How that type should look? Actually, I think it should be something like this:
```ts
defaultScale: number[] | string[];
```
Because if I'm not wrong, a scale can be a number array e.g. `[1, 4, 5]` or a string array: `['5px', '2rem']`

Reply:
A scale can also be nested object, for example:
```
theme = {
  colors: {
    gray: '#ccc',
    grays: [
      '#eee',
      '#ddd',
      '#333',
    ]
  }
}
```

Reply:
This is something I've been curious about as well, since I'm less familiar with TS. Since scales are defined in a user's own `theme` object, wouldn't it be ideal to define the scale types to match the schema that you're using in your theme? That might be easier said that done, but I'm curious of how something like that *could* be supported from Styled System side

Reply:
@jxnblk yeah this is the default theme, but can be extended on the user's side. Since the theme is dynamic is impossible for us to type check before hand.

On that note, maybe the Theme shouldn't be so flexible :) haha let's not go there.

Reply:
Yeah, I doubt this is something that can be tackled on the side of styled-system. Coincidentally I dug into this myself this week as we're currently adopting typescript and styled-system at the same time. If you go the custom type route you can do cool things like type the length & values of your scales with a tuple, i.e.

```
import { Theme } from 'styled-system';
import { ColorProperty } from 'csstype';

type CustomTheme = Theme & {
  colors: {
    red: [ColorProperty, ColorProperty, ColorProperty, ColorProperty, ColorProperty]
  }
};
```

But you'll have to implement your own types and you don't really have a way to easily feed these into the default styled-system setup. Plus with the amount of flexibility styled-system offers out of the box I doubt you'll ever be able to strictly type this. Also utils like `themeGet` and color props that take deeply nested string values are pretty much untypeable.

Reply:
> This is all pretty cool & interesting stuff but maybe it should be part of a separate discussion somewhere :)

Yeah, very cool discussion, but I agree that it doesn't fall under the scope of this PR. Unless there are any objections, let's leave in the `any` scale typing for now.